File: Maps/CountSubArray/cnt.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        if(n==0)return 0;
        int count = 0;
        unordered_map<int,int>mp;
        for(int i=1;i<n;i++){
            nums[i] += nums[i-1];
        }
        for(int i=0;i<n;i++){
            int ele = nums[i];
            if(ele == k)count++;
            int searchEle = ele-k;
            if(mp.find(searchEle)!=mp.end())count += mp[searchEle];
            mp[nums[i]]++;
        }
        return count;
    }
};

File: Maps/CountSubArray/cnt.py
class Solution:
    def cntSubarrays(self, arr, k):
        # code here
        n = len(arr)
        if n == 0: return 0
        for i in range(1,n):
            arr[i] += arr[i-1]
        count = 0
        mp = {}
        for i in range(n):
            ele = arr[i]
            if ele == k: count += 1
            searchEle = ele - k
            if searchEle in mp:
                count += mp[searchEle]
            mp[arr[i]] = mp.get(arr[i],0) + 1
        return count